### ISSUE TITLE
vscode.languages.getDiagnostics appears to return old diagnostics in some circumstances

### ISSUE THREAD

**Issue by lostintangent** [Microsoft/vscode#54359]
A user [reported an issue](https://github.com/MicrosoftDocs/live-share/issues/442#issuecomment-404366305) with VS Live Share, about diagnostics not being synchronized properly between participants in a session. After looking into it a bit, it appears like the issue may be with the `vscode.languages.getDiagnostics` method returning "stale" `Diagnostics`, depending on how the lifetime of the `DiagnosticCollection` is managed by extensions.

For example, take the following code (which is representative of what the [Elm](https://github.com/Krzysztof-Cieslak/vscode-elm) and Flow extensions are currently doing):

```javascript
vscode.workspace.onDidSaveTextDocument((d) => {
    let errors = vscode.languages.createDiagnosticCollection("foo");
    errors.clear(); 
    errors.set(d.uri, [new vscode.Diagnostic(new vscode.Range(0, 0, 0, 0), "Failure")]);

    const diagnostics = vscode.languages.getDiagnostics(d.uri);
    console.log(diagnostics);
}));
```

If you debug that, and edit/save a document a couple times, you'll see the following in the `Problems` pane, which is the expected behavior:

<img width="287" alt="screen shot 2018-07-15 at 3 58 55 pm" src="https://user-images.githubusercontent.com/116461/42739255-caeb0746-8848-11e8-8178-82f1eb5d19d5.png">

However, if you look in the `Debug Console`, you'll see the following printed, which is the output of the `languages.getDiagnostics` method:

<img width="176" alt="screen shot 2018-07-15 at 3 59 06 pm" src="https://user-images.githubusercontent.com/116461/42739260-f11b4534-8848-11e8-9ce5-87063cfa9d83.png">

So for some reason, the `Problems` pane is displaying the expected diagnostics. However, the `getDiagnostics` method is returning stale diagnostics. Ultimately, this results in errors never going away, since the result of calling `getDiagnostics` becomes purely additive. This is problematic for Live Share, since we use the results of the `getDiagnostics` method to sync diagnostics with guests.

If you modify the above code, to create a single `DiagnosticCollection` for the lifetime of the extension (as opposed to re-creating it over and over again), then things behave as expected:

```javascript
let errors = vscode.languages.createDiagnosticCollection("foo");
vscode.workspace.onDidSaveTextDocument((d) => {    
    errors.clear();
    errors.set(d.uri, [new vscode.Diagnostic(new vscode.Range(0, 0, 0, 0), "Failure")]);

    const diagnostics = vscode.languages.getDiagnostics(d.uri);
    console.log(diagnostics);
}));
```

Overall, the above code seems to make sense, and I [submitted a PR](https://github.com/Krzysztof-Cieslak/vscode-elm/pull/225) to the Elm extension to refactor their code as such. However, it would be ideal if the `vscode.languages.getDiagnostics` method was resilient to this situation, and returned the same `Diagnostics` that are being displayed in the `Problems` pane.

**Comment by lostintangent** [comment]
@jrieken Apologies. My original repro was too simple, and I'm not sure this is a matter of duplication. If I modify the diagnostic message like the following:

```javascript
let num = 0;

...

// Notice that I'm changing the error message each time to prevent de-duplication
errors.set(d.uri, [new vscode.Diagnostic(new vscode.Range(0, 0, 0, 0), `Failure${num++}`)]);
```

And save the document a few times, I see the following in the `Problems` pane:

<img width="261" alt="screen shot 2018-07-16 at 9 00 04 am" src="https://user-images.githubusercontent.com/116461/42769384-e2b1a6b2-88d6-11e8-83b3-69da8d97222a.png">

And the following output from calling `getDiagnostics`:

<img width="263" alt="screen shot 2018-07-16 at 8 59 57 am" src="https://user-images.githubusercontent.com/116461/42769394-eb383d00-88d6-11e8-9d9c-ac2f0f3b0015.png">

Are you implying that the `Problems` pane is somehow de-duplicating diagnostics based on some kind of pattern? In the above case, these diagnostics have different messages, and so I'm not sure why they would be removed from the UI.

Even further, if I call the `getDiagnostics()` method, instead of `getDiagnostics(uri)`, then the old diagnostics are gone! So this appears to be a bug with the `getDiagnostics(uri)` method specifically. Is it possible that the `Problems` pane is consuming the `getDiagnostics()` method (workspace-wide retrieval vs. document-specific), and that's why it's immune to this issue?

**Comment by jrieken** [comment]
>Even further, if I call the getDiagnostics() method, instead of getDiagnostics(uri), then the old diagnostics are gone!

Sure about that? Both methods loop over the same collection of DiagnosticsCollections... Be aware of the different return types. 

> So this appears to be a bug with the getDiagnostics(uri) method specifically. 

As long as you create more and more collections without disposing old ones, diagnostics pile up and the `getDiagnostics`-function returns them all. That's how it should be.

> Is it possible that the Problems pane is consuming the getDiagnostics() method (workspace-wide retrieval vs. document-specific), and that's why it's immune to this issue?

Fair question. Taking another look and turns out that the `name` you provide when creating a collection is actually used as some kind of identifier. That makes the old markers disappear (be overwritten). I'd say that's actually a bug and I will make a code accordingly, e.g. when using a name that's already being used, we will bring a warning and use a different internal identifier.

**Comment by lostintangent** [comment]
@jrieken Ah yeah, you're right. Both overloads of `getDiagnostics` have this same issue. Apologies for saying otherwise!

**Comment by dbaeumer** [comment]
Using a different owner for TypeScript and JavaScript works as long as users don't use a TSC task to check JS code. If they do the best we can do is to have different problem matchers with different IDs. I would suggest we simply make the change in the TS extension and then see if users get hit by tasks on pure JS projects (which I think is a very limited number of users in total).

**Comment by jrieken** [comment]
Ok. Alternatively, TypeScript can use one diagnostics collection (named typescript) for both languages

**Comment by jrieken** [comment]
I leave my change as it and leaving up to you two to fix this

**Comment by dbaeumer** [comment]
+1 for @jrieken suggestion since it is the right fix to do since JS and TS will not overlap.

**Comment by mjbvz** [comment]
Here's the change that unified the diagnostics in the first place: https://github.com/Microsoft/vscode/commit/b9459731b046fbb27e739e4822f0ca3894633e25

To fix #48709. It was done for exactly the reason you detailed: using tsc to check javascript code. 

Using a single collection will require work since we make some assumptions about having separate ones for JS and TS (it lets us handle the `typescript.validate` and `javascript.validate` settings more easily)

**Comment by dbaeumer** [comment]
@mjbvz thanks!
